From 007eda26153f6f1c7cb598f743671ed8912f8699 Mon Sep 17 00:00:00 2001
From: Kiveisha Yevgeniy <yevgeniy.kiveisha@intel.com>
Date: Sun, 11 Jan 2015 19:14:14 +0200
Subject: [PATCH 1/2] camera ov5693 and mt9m114: device power_up retries.

Added __power_up function which called from power_up and
incase of failure it will be called again.
Number or retries limited to 4. Added to both cameras.

Issue: GMINL-3806
Change-Id: Ib300182038476fada3c1fb504d16cddb766254a6
Signed-off-by: Kiveisha Yevgeniy <yevgeniy.kiveisha@intel.com>
---
 .../camera/drivers/media/i2c/mt9m114.c                 | 17 ++++++++++++++++-
 .../camera/drivers/media/i2c/ov5693-ecs/ov5693.c       | 18 ++++++++++++++++--
 2 files changed, 32 insertions(+), 3 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/mt9m114.c b/drivers/external_drivers/camera/drivers/media/i2c/mt9m114.c
index 22c6b7a..a98c102 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/mt9m114.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/mt9m114.c
@@ -536,7 +536,7 @@ static int gpio_ctrl(struct v4l2_subdev *sd, bool flag)
 	return ret;
 }
 
-static int power_up(struct v4l2_subdev *sd)
+static int __power_up(struct v4l2_subdev *sd)
 {
 	struct mt9m114_device *dev = to_mt9m114_sensor(sd);
 	struct i2c_client *client = v4l2_get_subdevdata(sd);
@@ -609,6 +609,21 @@ static int power_down(struct v4l2_subdev *sd)
 	return ret;
 }
 
+static int power_up(struct v4l2_subdev *sd)
+{
+	static const int retry_count = 4;
+	int i, ret;
+
+	for (i = 0; i < retry_count; i++) {
+		ret = __power_up(sd);
+		if (!ret)
+			return 0;
+
+		power_down(sd);
+	}
+	return ret;
+}
+
 static int mt9m114_s_power(struct v4l2_subdev *sd, int power)
 {
 	if (power == 0)
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
index 6745f2a..e0fc8fc 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
@@ -1334,8 +1334,7 @@ static int gpio_ctrl(struct v4l2_subdev *sd, bool flag)
 	return ret;
 }
 
-
-static int power_up(struct v4l2_subdev *sd)
+static int __power_up(struct v4l2_subdev *sd)
 {
 	struct ov5693_device *dev = to_ov5693_sensor(sd);
 	struct i2c_client *client = v4l2_get_subdevdata(sd);
@@ -1415,6 +1414,21 @@ static int power_down(struct v4l2_subdev *sd)
 	return ret;
 }
 
+static int power_up(struct v4l2_subdev *sd)
+{
+	static const int retry_count = 4;
+	int i, ret;
+
+	for (i = 0; i < retry_count; i++) {
+		ret = __power_up(sd);
+		if (!ret)
+			return 0;
+
+		power_down(sd);
+	}
+	return ret;
+}
+
 static int ov5693_s_power(struct v4l2_subdev *sd, int on)
 {
 	int ret;
-- 
1.9.1

